src/main/scala/calc/Compiler.scala

```diff
@@ -74,6 +85,74 @@ object Compiler {
       case Literal(value) =>
         irt.DoubleLiteral(value)
 
+      case BinaryOp(op, rhs, lhs) =>
+        val right = compileExpr(rhs, listId)
+        val left  = compileExpr(lhs, listId)
+        irt.BinaryOp(binaryOpMap(op), right, left)
+
+      case Let(name, value, body) =>
+        val irtId = irt.Ident(name.name)
+        val letVal = compileExpr(value, listId)
+        val letTree = irt.Block(List(
+          irt.VarDef(irtId, letVal.tpe, false, letVal),
+          compileExpr(body, listId + (name.name -> letVal.tpe)))
+        )
+        nativeFun.contains(name.toString) match{
+          case true  => throw new Exception(s"${name.name} can't be redefined")
+          case false => letTree
```

Comment:
This is an arbitrary restriction. It should be possible to shadow `sin` from the global scope, like in `let sin = fun(x) = { x + 1} in sin(3)` or even `let sin = 3 in sin`.

You can tell whether a given math function has been shadowed by testing whether `listId.contains(name)`